_drafts/ble-throughput-primer.md

```diff
@@ -0,0 +1,288 @@
+
+### Implementation Specific Details to Check
+
+{:.no_toc}
+
+#### Supported Packets Per Connection Event
+
+The number of packets which can be transmitted in one _connection event_ is limited by the Bluetooth software stack in use. Some older devices only support one packet per connection interval. Assuming the fastest connection interval, _7.5ms_, this would mean the max raw data throughput which could be realized is only:
+
+`27 bytes / 7.5ms = 3.6kBytes / second = 0.029 Mbps`
+
+The end of a connection event is controlled by the _More Data_ (_MD_) bit in the header of the link layer packet discussed [above](#ll-packet). The LL packet header looks like this:
+
+![](img/ble-throughput/ll-pdu-header.png)
+
+A good description of how the MD bit works can be found in the Bluetooth Core Specification [^4]
+
+![](img/ble-throughput/ll-md-bit-desc.png)
+
+If you have a protocol analyzer, you can analyze how the MD is being set to figure out what side of the connection is responsible for limiting the throughput.
+
+{:.no_toc}
+
+#### Supported Connection Intervals
+
+Many devices only support a subset of the valid connection parameter range (7.5ms - 4s). For example, Apple documents this information quite well in Section "25.6 Connection Parameters" of the "Accessory Design Guidelines For Apple Devices" [^7]:
+
+> ![](img/ble-throughput/apple-connection-parameters.png)
+
+A longer connection event in itself isn't necessarily a problem for throughput. Some devices will send and recieve data for an entire connection event so if you are streaming data continuously the same throughput can be realized even if you are using a 30ms interval instead of a 15ms interval.
+
+What the connection interval _does_ impact is latency. Many higher layer protocols have messages which need to be acknowledged by higher level software components. The longer the connection interval, the longer it will take for these round trips to complete
+
+{:.no_toc}
+
+#### Maximum Supported ATT MTU Size
+
+As we discussed [above](#max-throughput-gatt), the ATT MTU size has an impact on the max throughput which can be realized when using GATT due to protocol overhead. Some chips only support the default size (especially older Android Phones). When first testing a chip it's important to check and make sure larger packets can be sent!
```

Comment:
Maybe mention to also look at queue sizes for outbound GATT notifications and inbound writes w/o responses? I'm not sure which one it was, but I've seen a stack where you could only have one pending outbound notification... Also, not all stacks have events/cbs for queue full and drained.